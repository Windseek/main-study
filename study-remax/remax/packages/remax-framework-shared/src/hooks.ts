import { useLayoutEffect, useContext } from 'react';
import { Callback, lifeCycleName, registerLifecycle } from './lifecycle';
import PageInstanceContext from './PageInstanceContext';
import ComponentInstanceContext from './ComponentInstanceContext';
import AppInstanceContext from './AppInstanceContext';

export function usePageEvent(eventName: string, callback: Callback) {
  const pageInstance = useContext(PageInstanceContext);
  const lifeCycle = lifeCycleName(eventName);
  useLayoutEffect(() => {
    return registerLifecycle(pageInstance, lifeCycle, callback);
  });
}

export function useComponentInstance() {
  return useContext(ComponentInstanceContext);
}

export function usePageInstance() {
  return useContext(PageInstanceContext);
}

/**
 * App Hooks
 */

export function useAppEvent(eventName: string, callback: Callback) {
  const lifeCycle = lifeCycleName(eventName);
  useLayoutEffect(() => {
    return registerLifecycle(AppInstanceContext, lifeCycle, callback);
  });
}
