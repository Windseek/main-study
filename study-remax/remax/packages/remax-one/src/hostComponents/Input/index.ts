import * as React from 'react';
import memoizeOne from 'memoize-one';
import { InputEvent } from '../../types';
import { createCallback, createInputEvent, aliasProps } from '../../createHostComponent';
import type { InputProps } from './props';
import alias from './props/alias';

export type { InputProps };

export interface InputState {
  value?: string;
  controlled: boolean;
}

const createInputCallback = memoizeOne(createCallback);
const createConfirmCallback = memoizeOne(createCallback);
const createFocusCallback = memoizeOne(createCallback);
const createBlurCallback = memoizeOne(createCallback);

export default class Input extends React.Component<InputProps, InputState> {
  // 平台独有的属性默认值写在这
  static defaultProps = {
    'toutiao-selection-end': 999,
    'toutiao-selection-start': 999,

    'wechat-placeholder-class': 'input-placeholder',
    'wechat-cursor-spacing': 0,
    'wechat-confirm-type': 'done',
    'wechat-confirm-hold': false,
    'wechat-selection-end': -1,
    'wechat-selection-start': -1,
    'wechat-adjust-position': true,
    'wechat-hold-keyboard': false,
  };

  state: InputState = {
    value: '',
    // 阿里
    controlled: false,
  };

  constructor(props: InputProps) {
    super(props);

    const controlled = props.value !== undefined;
    const value = controlled ? props.value : props.defaultValue;

    this.state = {
      value,
      controlled,
    };
  }

  componentDidUpdate() {
    if (this.props.value !== undefined && this.props.value !== this.state.value) {
      this.setState({ value: this.props.value });
    }
  }

  handleInput = (e: InputEvent) => {
    const { controlled } = this.state;

    if (typeof this.props.onInput === 'function') {
      return this.props.onInput(e);
    }

    // 微信
    // 注意，微信要对 input 受控，必须要在 onInput 方法返回值
    return controlled ? this.props.value : undefined;
  };

  render() {
    const inputProps = { ...this.props };

    if (inputProps.onInput) {
      inputProps.onInput = createInputCallback(this.handleInput, createInputEvent);
    }

    if (inputProps.onConfirm) {
      inputProps.onConfirm = createConfirmCallback(this.props.onConfirm, createInputEvent);
    }

    if (inputProps.onFocus) {
      inputProps.onFocus = createFocusCallback(this.props.onFocus, createInputEvent);
    }

    if (inputProps.onBlur) {
      inputProps.onBlur = createBlurCallback(this.props.onBlur, createInputEvent);
    }

    // 通用属性的默认属性根据平台在这里设置
    if (process.env.REMAX_PLATFORM === 'toutiao') {
      inputProps.maxLength = inputProps.maxLength ?? 140;
    }

    if (process.env.REMAX_PLATFORM === 'wechat') {
      inputProps.type = inputProps.type ?? 'text';
      inputProps.password = inputProps.password ?? false;
      inputProps.disabled = inputProps.disabled ?? false;
      inputProps.focus = inputProps.focus ?? false;
      inputProps.maxLength = inputProps.maxLength ?? 140;
    }

    const nextProps = aliasProps(
      {
        ...inputProps,
        ...this.state,
      },
      alias
    );

    return React.createElement('input', nextProps);
  }
}
