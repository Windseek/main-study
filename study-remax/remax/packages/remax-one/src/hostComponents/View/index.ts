import * as React from 'react';
import createHostComponent from '../../createHostComponent';
import ViewProps from './props';
import defaults from './props/default';

export type { ViewProps };

const View: React.ComponentType<ViewProps> = createHostComponent<ViewProps>('view', null, defaults);

export default View;
