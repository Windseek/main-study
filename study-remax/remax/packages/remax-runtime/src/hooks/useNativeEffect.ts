import { useLayoutEffect, DependencyList } from 'react';
import nativeEffect, { Listener } from '../nativeEffect';

export default function useNativeEffect(listener: Listener, deps?: DependencyList) {
  useLayoutEffect(() => {
    return nativeEffect.connect(listener, !!deps);
  }, deps);
}
