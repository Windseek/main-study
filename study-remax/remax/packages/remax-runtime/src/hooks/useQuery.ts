import { useContext } from 'react';
import { PageInstanceContext } from '@remax/framework-shared';

export default function useQuery<
  Q extends Record<string, string | undefined> = { [name: string]: string | undefined }
>(): Q {
  const pageInstance: any = useContext(PageInstanceContext);
  return pageInstance.query;
}
