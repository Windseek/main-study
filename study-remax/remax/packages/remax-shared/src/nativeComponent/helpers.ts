import * as fs from 'fs';
import path from 'path';
import { slash } from '../path';

export const isNativeComponent = (sourcePath: string | null): boolean => {
  if (!sourcePath) {
    return false;
  }

  if (path.extname(sourcePath) !== '.js') {
    return false;
  }

  const sourceJsonPath = sourcePath.replace(/\.js$/, '.json');

  if (!fs.existsSync(sourceJsonPath)) {
    return false;
  }

  return require(sourceJsonPath).component;
};

export const getPath = (from: string, to: string) => {
  return slash(path.resolve(path.dirname(from), to));
};

const NATIVE_COMPONENT_OUTPUT_DIR = 'remaxVendors';

export function getNativeEntryAssetOutputPath(sourcePath: string, options: { cwd: string; rootDir: string }) {
  return (
    NATIVE_COMPONENT_OUTPUT_DIR +
    '/' +
    slash(sourcePath)
      .replace(slash(options.cwd) + '/', '')
      .replace(slash(options.rootDir) + '/', '')
      .replace(/@/g, '_')
      .replace(/node_modules/g, 'npm')
  );
}
