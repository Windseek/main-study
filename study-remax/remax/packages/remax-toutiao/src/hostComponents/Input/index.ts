import * as React from 'react';

import { BaseProps } from '../../types/component';
import { createHostComponent } from '@remax/runtime';

export interface InputProps extends BaseProps {
  name?: string;
  focus?: boolean;
  maxlength?: number;
  value?: string;
  password?: boolean;
  type?: 'text' | 'number' | 'digit';
  cursor?: number;
  selectionStart?: number;
  selectionEnd?: number;
  adjustPosition?: boolean;
  placeholder?: string;
  placeholderStyle?: React.CSSProperties;
  placeholderClassName?: string;
  disabled?: boolean;
  cursorSpacing?: number;
  onInput?: (e: any) => void | string;
  onClick?: (e: any) => void;
  onFocus?: (e: any) => void;
  onBlur?: (e: any) => void;
  onConfirm?: (e: any) => void;
}

export const Input: React.ComponentType<InputProps> = createHostComponent<InputProps>('input');

Input.defaultProps = {
  type: 'text',
  password: false,
  value: '',
  disabled: false,
  maxlength: 140,
  focus: false,
  selectionEnd: -1,
  selectionStart: -1,
  cursorSpacing: 0,
  cursor: -1,
};
