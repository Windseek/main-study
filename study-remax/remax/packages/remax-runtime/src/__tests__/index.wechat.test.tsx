import * as React from 'react';
import './helpers/setupGlobals';
import View from './helpers/View';
import Input from './helpers/Input';
import render from '../render';
import { reset as resetInstanceId } from '../instanceId';
import Container from '../Container';
import { useNativeEffect } from '../hooks';

function delay(ms: number): Promise<void> {
  if (typeof ms !== 'number') {
    throw new Error('Must specify ms');
  }
  return new Promise(resolve => {
    setTimeout(() => {
      resolve();
    }, ms);
  });
}

const p = {
  setData(state: any, callback: () => void) {
    setTimeout(() => {
      if (typeof callback === 'function') {
        callback();
      }
    });
  },
};

describe('wechat remax render', () => {
  afterEach(() => {
    resetInstanceId();
  });

  it('render correctly', () => {
    const Page = () => <View className="foo">hello</View>;
    const container = new Container(p);
    render(<Page />, container);
    expect(container.root).toMatchSnapshot();
  });

  it('insert new element', () => {
    class Page extends React.Component {
      state = {
        list: [1, 3],
      };

      update() {
        this.setState({
          list: [1, 2, 3],
        });
      }

      render() {
        const { list } = this.state;
        return (
          <View>
            {list.map(i => (
              <View key={i}>{i}</View>
            ))}
          </View>
        );
      }
    }

    const container = new Container(p);
    const page = React.createRef<any>();
    render(<Page ref={page} />, container);
    expect(container.root).toMatchSnapshot();
    page.current.update();
    expect(container.root).toMatchSnapshot();
  });

  it("change elements' order", () => {
    class Page extends React.Component {
      state = {
        list: [1, 2, 3],
      };

      updateA() {
        this.setState({
          list: [2, 1, 3],
        });
      }

      updateB() {
        this.setState({
          list: [2, 3, 1],
        });
      }

      render() {
        const { list } = this.state;
        return (
          <View>
            {list.map(i => (
              <View key={i}>{i}</View>
            ))}
          </View>
        );
      }
    }

    const container = new Container(p);
    const page = React.createRef<any>();
    render(<Page ref={page} />, container);
    expect(container.root).toMatchSnapshot();
    page.current.updateA();
    expect(container.root).toMatchSnapshot();
    page.current.updateB();
    expect(container.root).toMatchSnapshot();
  });

  it('umount component', () => {
    class Page extends React.Component {
      state = {
        show: true,
      };

      hide() {
        this.setState({ show: false });
      }

      render() {
        return <View>{this.state.show && <View>foo</View>}</View>;
      }
    }
    const container = new Container(p);
    const page = React.createRef<any>();
    render(<Page ref={page} />, container);
    expect(container.root).toMatchSnapshot();
    page.current.hide();
    expect(container.root).toMatchSnapshot();
  });

  it('renders style', () => {
    const Page = () => <View style={{ width: '100px', height: '100px' }}>hello</View>;
    const container = new Container(p);
    render(<Page />, container);
    expect(container.root).toMatchSnapshot();
  });

  it('renders vendor prefix style', () => {
    const Page = () => <View style={{ WebkitLineClamp: 2, height: '100px' }}>hello</View>;
    const container = new Container(p);
    render(<Page />, container);
    expect(container.root).toMatchSnapshot();
  });

  it('renders empty style', () => {
    const Page = () => <View style={undefined}>hello</View>;
    const container = new Container(p);
    render(<Page />, container);
    expect(container.root).toMatchSnapshot();
  });

  it('renders conditional fragment correctly', () => {
    class Page extends React.Component {
      state = {
        show: false,
      };

      show() {
        this.setState({ show: true });
      }

      render() {
        if (this.state.show) {
          return <View>foo</View>;
        }

        return (
          <>
            <View>one</View>
            <View>two</View>
          </>
        );
      }
    }
    const container = new Container(p);
    const page = React.createRef<any>();
    render(<Page ref={page} />, container);
    expect(container.root).toMatchSnapshot();
    page.current.show();
    expect(container.root).toMatchSnapshot();
  });

  it('render native component correctly', () => {
    const NativeComponent = ({ fooBar, onClick, className }: any) =>
      React.createElement('native-component', {
        fooBar,
        className,
        onClick,
      });
    const actions: any = [];
    const p = {
      setData: (data: any) => actions.push(data),
    };

    const container = new Container(p);
    render(<NativeComponent fooBar="fooBar" onClick={() => void 0} className="class" />, container);

    expect(actions).toMatchSnapshot();
  });

  it('remove and add children', async () => {
    const logs: any = [];
    class App extends React.Component {
      state = {
        toggle: false,
      };

      toggle() {
        this.setState({
          toggle: true,
        });
      }

      render() {
        const { toggle } = this.state;
        return (
          <>
            <View id="1">
              <View id="2">{toggle ? <View id="3">a</View> : <View id="3">b</View>}</View>
              {!toggle && <View id="5">safeAreaHeight</View>}
            </View>
          </>
        );
      }
    }

    const container = new Container({
      setData(action: any) {
        logs.push(action);
      },
    });
    const app = React.createRef<any>();
    render(<App ref={app} />, container);
    app.current.toggle();
    await delay(100);
    expect(logs.pop()).toEqual({
      'root.nodes.6.children': [3],
      'root.nodes.6.nodes.3.nodes.2.nodes.1': {
        id: 1,
        text: 'a',
        type: 'plain-text',
      },
      'root.nodes.6.nodes.5': null,
    });
  });

  it('swaps items', () => {
    const logs: any = [];
    class App extends React.Component {
      state = {
        list: ['a', 'b'],
      };

      swap() {
        const { list } = this.state;
        this.setState({
          list: [list[1], list[0]],
        });
      }

      render() {
        const { list } = this.state;
        return (
          <View id="1">
            {list.map((item, index) => (
              <View key={item}>
                {index}
                {item}
              </View>
            ))}
          </View>
        );
      }
    }

    const container = new Container({
      setData(action: any) {
        logs.push(action);
      },
    });
    const app = React.createRef<any>();
    render(<App ref={app} />, container);
    expect(logs.pop()).toMatchSnapshot();
    app.current.swap();
    expect(logs.pop()).toMatchSnapshot();
    app.current.swap();
    expect(logs.pop()).toMatchSnapshot();
  });

  it("does not update deleted node's children node", async () => {
    const logs: any = [];
    class App extends React.Component {
      state = {
        a: true,
        b: true,
      };

      toggle() {
        this.setState({
          b: false,
        });
        this.setState({
          a: false,
        });
      }

      render() {
        const { a, b } = this.state;
        // eslint-disable-next-line prettier/prettier
        return <View id="a">{a && <View id="b">{b ? <View id="c">a</View> : 'b'}</View>}</View>;
      }
    }

    const container = new Container({
      setData(action: any) {
        logs.push(action);
      },
    });
    const app = React.createRef<any>();
    render(<App ref={app} />, container);
    app.current.toggle();
    await delay(100);
    expect(logs.pop()).toEqual({
      'root.nodes.4.children': [],
      'root.nodes.4.nodes.3': null,
    });
  });

  it('does not update deleted node', async () => {
    const logs: any = [];
    class App extends React.Component {
      state = {
        a: true,
        b: true,
      };

      toggle() {
        this.setState({
          b: false,
        });
        this.setState({
          a: false,
        });
      }

      render() {
        const { a, b } = this.state;
        return <View>{a && <View style={{ color: b ? 'red' : 'green' }}>foo</View>}</View>;
      }
    }

    const container = new Container({
      setData(action: any) {
        logs.push(action);
      },
    });
    const app = React.createRef<any>();
    render(<App ref={app} />, container);
    app.current.toggle();
    await delay(100);
    expect(logs.pop()).toEqual({
      'root.nodes.3.children': [],
      'root.nodes.3.nodes.2': null,
    });
  });

  it('create proxy for onClick callback', () => {
    const view = React.createRef<any>();
    const handleClick = () => void 0;
    const handleAnimationStart = () => void 0;
    class Page extends React.Component {
      render() {
        return <View ref={view} onClick={handleClick} onAnimationStart={handleAnimationStart} />;
      }
    }
    const container = new Container(p);
    render(<Page />, container);

    function findFn(name: string) {
      const fnKeys = Object.keys(view.current.container.context);
      const fnKey = fnKeys.find(key => key.indexOf(name) !== -1) || '';
      return view.current.container.context[fnKey];
    }

    const newHandleClick = findFn('onTap');
    const newHandleAnimationStart = findFn('onAnimationStart');

    expect(newHandleClick).not.toBe(handleClick);
    expect(newHandleAnimationStart).toBe(handleAnimationStart);
  });

  it('useEffect works', done => {
    const Page = () => {
      React.useEffect(() => {
        done();
      });

      return <View>app</View>;
    };
    const container = new Container(p);
    render(<Page />, container);
  });

  it('pure rerender when props changed', done => {
    const payload: any[] = [];
    const context = {
      setData: (data: any) => {
        payload.push(data);
      },
    };

    class Page extends React.Component {
      state = {
        value: 'foo',
      };

      setValue(value: string) {
        this.setState({ value });
      }

      render() {
        return (
          <View style={{ width: '32px' }}>
            <Input value={this.state.value} />
          </View>
        );
      }
    }
    const container = new Container(context);
    const page = React.createRef<any>();
    render(<Page ref={page} />, container);

    expect.assertions(2);

    page.current.setValue('bar');

    setTimeout(() => {
      expect(payload).toHaveLength(2);
      expect(payload[1]).toMatchInlineSnapshot(`
        Object {
          "root.nodes.2.nodes.1.props.value": "bar",
        }
      `);
      done();
    }, 5);
  });

  it('pure rerender when props delete', done => {
    const payload: any[] = [];
    const context = {
      setData: (data: any) => {
        payload.push(data);
      },
    };

    class Page extends React.Component {
      state = {
        value: 'foo',
      };

      setValue(value: string) {
        this.setState({ value });
      }

      render() {
        return (
          <View style={{ width: '32px' }}>{!this.state.value ? <Input /> : <Input value={this.state.value} />}</View>
        );
      }
    }
    const container = new Container(context);
    const page = React.createRef<any>();
    render(<Page ref={page} />, container);

    expect.assertions(2);

    page.current.setValue(undefined);

    setTimeout(() => {
      expect(payload).toHaveLength(2);
      expect(payload).toMatchInlineSnapshot(`
        Array [
          Object {
            "root.children": Array [
              2,
            ],
            "root.nodes.2": Object {
              "children": Array [
                1,
              ],
              "id": 2,
              "nodes": Object {
                "1": Object {
                  "children": Array [],
                  "id": 1,
                  "props": Object {
                    "value": "foo",
                  },
                  "text": undefined,
                  "type": "input",
                },
              },
              "props": Object {
                "style": "width:32rpx;",
              },
              "text": undefined,
              "type": "view",
            },
          },
          Object {
            "root.nodes.2.nodes.1.props.value": null,
          },
        ]
      `);
      done();
    }, 5);
  });

  it('useNativeEffect once works', done => {
    let count = 0;
    const Page = () => {
      const [width, setWidth] = React.useState(0);
      useNativeEffect(() => {
        count += 1;

        setTimeout(() => {
          if (count === 1) {
            done();
          }
        }, 500);
      }, []);
      React.useEffect(() => {
        setTimeout(() => {
          setWidth(100);
        }, 100);
      }, []);

      return <View>{width}</View>;
    };
    const container = new Container(p);
    render(<Page />, container);
  });

  it('useNativeEffect deps works', done => {
    let count = 0;
    const Page = () => {
      const [width, setWidth] = React.useState(0);
      const [height, setheight] = React.useState(0);
      useNativeEffect(() => {
        count += 1;

        if (count === 2) {
          done();
        }
      }, [width]);
      React.useEffect(() => {
        setheight(100);
        setTimeout(() => {
          setWidth(100);
        }, 1000);
      }, []);

      return (
        <View>
          {width}
          {height}
        </View>
      );
    };
    const container = new Container(p);
    render(<Page />, container);
  });
});
