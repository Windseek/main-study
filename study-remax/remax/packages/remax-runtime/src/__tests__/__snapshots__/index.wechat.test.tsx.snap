// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wechat remax render change elements' order 1`] = `
Object {
  "children": Array [
    7,
  ],
  "id": 0,
  "nodes": Object {
    "7": Object {
      "children": Array [
        2,
        4,
        6,
      ],
      "id": 7,
      "nodes": Object {
        "2": Object {
          "children": Array [
            1,
          ],
          "id": 2,
          "nodes": Object {
            "1": Object {
              "id": 1,
              "text": "1",
              "type": "plain-text",
            },
          },
          "props": Object {},
          "text": undefined,
          "type": "view",
        },
        "4": Object {
          "children": Array [
            3,
          ],
          "id": 4,
          "nodes": Object {
            "3": Object {
              "id": 3,
              "text": "2",
              "type": "plain-text",
            },
          },
          "props": Object {},
          "text": undefined,
          "type": "view",
        },
        "6": Object {
          "children": Array [
            5,
          ],
          "id": 6,
          "nodes": Object {
            "5": Object {
              "id": 5,
              "text": "3",
              "type": "plain-text",
            },
          },
          "props": Object {},
          "text": undefined,
          "type": "view",
        },
      },
      "props": Object {},
      "text": undefined,
      "type": "view",
    },
  },
  "props": undefined,
  "text": undefined,
  "type": "root",
}
`;

exports[`wechat remax render change elements' order 2`] = `
Object {
  "children": Array [
    7,
  ],
  "id": 0,
  "nodes": Object {
    "7": Object {
      "children": Array [
        4,
        2,
        6,
      ],
      "id": 7,
      "nodes": Object {
        "2": Object {
          "children": Array [
            1,
          ],
          "id": 2,
          "nodes": Object {
            "1": Object {
              "id": 1,
              "text": "1",
              "type": "plain-text",
            },
          },
          "props": Object {},
          "text": undefined,
          "type": "view",
        },
        "4": Object {
          "children": Array [
            3,
          ],
          "id": 4,
          "nodes": Object {
            "3": Object {
              "id": 3,
              "text": "2",
              "type": "plain-text",
            },
          },
          "props": Object {},
          "text": undefined,
          "type": "view",
        },
        "6": Object {
          "children": Array [
            5,
          ],
          "id": 6,
          "nodes": Object {
            "5": Object {
              "id": 5,
              "text": "3",
              "type": "plain-text",
            },
          },
          "props": Object {},
          "text": undefined,
          "type": "view",
        },
      },
      "props": Object {},
      "text": undefined,
      "type": "view",
    },
  },
  "props": undefined,
  "text": undefined,
  "type": "root",
}
`;

exports[`wechat remax render change elements' order 3`] = `
Object {
  "children": Array [
    7,
  ],
  "id": 0,
  "nodes": Object {
    "7": Object {
      "children": Array [
        4,
        6,
        2,
      ],
      "id": 7,
      "nodes": Object {
        "2": Object {
          "children": Array [
            1,
          ],
          "id": 2,
          "nodes": Object {
            "1": Object {
              "id": 1,
              "text": "1",
              "type": "plain-text",
            },
          },
          "props": Object {},
          "text": undefined,
          "type": "view",
        },
        "4": Object {
          "children": Array [
            3,
          ],
          "id": 4,
          "nodes": Object {
            "3": Object {
              "id": 3,
              "text": "2",
              "type": "plain-text",
            },
          },
          "props": Object {},
          "text": undefined,
          "type": "view",
        },
        "6": Object {
          "children": Array [
            5,
          ],
          "id": 6,
          "nodes": Object {
            "5": Object {
              "id": 5,
              "text": "3",
              "type": "plain-text",
            },
          },
          "props": Object {},
          "text": undefined,
          "type": "view",
        },
      },
      "props": Object {},
      "text": undefined,
      "type": "view",
    },
  },
  "props": undefined,
  "text": undefined,
  "type": "root",
}
`;

exports[`wechat remax render insert new element 1`] = `
Object {
  "children": Array [
    5,
  ],
  "id": 0,
  "nodes": Object {
    "5": Object {
      "children": Array [
        2,
        4,
      ],
      "id": 5,
      "nodes": Object {
        "2": Object {
          "children": Array [
            1,
          ],
          "id": 2,
          "nodes": Object {
            "1": Object {
              "id": 1,
              "text": "1",
              "type": "plain-text",
            },
          },
          "props": Object {},
          "text": undefined,
          "type": "view",
        },
        "4": Object {
          "children": Array [
            3,
          ],
          "id": 4,
          "nodes": Object {
            "3": Object {
              "id": 3,
              "text": "3",
              "type": "plain-text",
            },
          },
          "props": Object {},
          "text": undefined,
          "type": "view",
        },
      },
      "props": Object {},
      "text": undefined,
      "type": "view",
    },
  },
  "props": undefined,
  "text": undefined,
  "type": "root",
}
`;

exports[`wechat remax render insert new element 2`] = `
Object {
  "children": Array [
    5,
  ],
  "id": 0,
  "nodes": Object {
    "5": Object {
      "children": Array [
        2,
        7,
        4,
      ],
      "id": 5,
      "nodes": Object {
        "2": Object {
          "children": Array [
            1,
          ],
          "id": 2,
          "nodes": Object {
            "1": Object {
              "id": 1,
              "text": "1",
              "type": "plain-text",
            },
          },
          "props": Object {},
          "text": undefined,
          "type": "view",
        },
        "4": Object {
          "children": Array [
            3,
          ],
          "id": 4,
          "nodes": Object {
            "3": Object {
              "id": 3,
              "text": "3",
              "type": "plain-text",
            },
          },
          "props": Object {},
          "text": undefined,
          "type": "view",
        },
        "7": Object {
          "children": Array [
            6,
          ],
          "id": 7,
          "nodes": Object {
            "6": Object {
              "id": 6,
              "text": "2",
              "type": "plain-text",
            },
          },
          "props": Object {},
          "text": undefined,
          "type": "view",
        },
      },
      "props": Object {},
      "text": undefined,
      "type": "view",
    },
  },
  "props": undefined,
  "text": undefined,
  "type": "root",
}
`;

exports[`wechat remax render render correctly 1`] = `
Object {
  "children": Array [
    2,
  ],
  "id": 0,
  "nodes": Object {
    "2": Object {
      "children": Array [
        1,
      ],
      "id": 2,
      "nodes": Object {
        "1": Object {
          "id": 1,
          "text": "hello",
          "type": "plain-text",
        },
      },
      "props": Object {
        "className": "foo",
      },
      "text": undefined,
      "type": "view",
    },
  },
  "props": undefined,
  "text": undefined,
  "type": "root",
}
`;

exports[`wechat remax render render native component correctly 1`] = `
Array [
  Object {
    "root.children": Array [
      1,
    ],
    "root.nodes.1": Object {
      "children": Array [],
      "id": 1,
      "props": Object {
        "className": "class",
        "fooBar": "fooBar",
        "onClick": "$$REMAX_METHOD_1_onClick",
      },
      "text": undefined,
      "type": "native-component",
    },
  },
]
`;

exports[`wechat remax render renders conditional fragment correctly 1`] = `
Object {
  "children": Array [
    2,
    4,
  ],
  "id": 0,
  "nodes": Object {
    "2": Object {
      "children": Array [
        1,
      ],
      "id": 2,
      "nodes": Object {
        "1": Object {
          "id": 1,
          "text": "one",
          "type": "plain-text",
        },
      },
      "props": Object {},
      "text": undefined,
      "type": "view",
    },
    "4": Object {
      "children": Array [
        3,
      ],
      "id": 4,
      "nodes": Object {
        "3": Object {
          "id": 3,
          "text": "two",
          "type": "plain-text",
        },
      },
      "props": Object {},
      "text": undefined,
      "type": "view",
    },
  },
  "props": undefined,
  "text": undefined,
  "type": "root",
}
`;

exports[`wechat remax render renders conditional fragment correctly 2`] = `
Object {
  "children": Array [
    2,
  ],
  "id": 0,
  "nodes": Object {
    "2": Object {
      "children": Array [
        1,
      ],
      "id": 2,
      "nodes": Object {
        "1": Object {
          "id": 1,
          "text": "foo",
          "type": "plain-text",
        },
      },
      "props": Object {},
      "text": undefined,
      "type": "view",
    },
  },
  "props": undefined,
  "text": undefined,
  "type": "root",
}
`;

exports[`wechat remax render renders empty style 1`] = `
Object {
  "children": Array [
    2,
  ],
  "id": 0,
  "nodes": Object {
    "2": Object {
      "children": Array [
        1,
      ],
      "id": 2,
      "nodes": Object {
        "1": Object {
          "id": 1,
          "text": "hello",
          "type": "plain-text",
        },
      },
      "props": Object {
        "style": "",
      },
      "text": undefined,
      "type": "view",
    },
  },
  "props": undefined,
  "text": undefined,
  "type": "root",
}
`;

exports[`wechat remax render renders style 1`] = `
Object {
  "children": Array [
    2,
  ],
  "id": 0,
  "nodes": Object {
    "2": Object {
      "children": Array [
        1,
      ],
      "id": 2,
      "nodes": Object {
        "1": Object {
          "id": 1,
          "text": "hello",
          "type": "plain-text",
        },
      },
      "props": Object {
        "style": "width:100rpx;height:100rpx;",
      },
      "text": undefined,
      "type": "view",
    },
  },
  "props": undefined,
  "text": undefined,
  "type": "root",
}
`;

exports[`wechat remax render renders vendor prefix style 1`] = `
Object {
  "children": Array [
    2,
  ],
  "id": 0,
  "nodes": Object {
    "2": Object {
      "children": Array [
        1,
      ],
      "id": 2,
      "nodes": Object {
        "1": Object {
          "id": 1,
          "text": "hello",
          "type": "plain-text",
        },
      },
      "props": Object {
        "style": "-webkit-line-clamp:2;height:100rpx;",
      },
      "text": undefined,
      "type": "view",
    },
  },
  "props": undefined,
  "text": undefined,
  "type": "root",
}
`;

exports[`wechat remax render swaps items 1`] = `
Object {
  "root.children": Array [
    7,
  ],
  "root.nodes.7": Object {
    "children": Array [
      3,
      6,
    ],
    "id": 7,
    "nodes": Object {
      "3": Object {
        "children": Array [
          1,
          2,
        ],
        "id": 3,
        "nodes": Object {
          "1": Object {
            "id": 1,
            "text": "0",
            "type": "plain-text",
          },
          "2": Object {
            "id": 2,
            "text": "a",
            "type": "plain-text",
          },
        },
        "props": Object {},
        "text": undefined,
        "type": "view",
      },
      "6": Object {
        "children": Array [
          4,
          5,
        ],
        "id": 6,
        "nodes": Object {
          "4": Object {
            "id": 4,
            "text": "1",
            "type": "plain-text",
          },
          "5": Object {
            "id": 5,
            "text": "b",
            "type": "plain-text",
          },
        },
        "props": Object {},
        "text": undefined,
        "type": "view",
      },
    },
    "props": Object {
      "id": "1",
    },
    "text": undefined,
    "type": "view",
  },
}
`;

exports[`wechat remax render swaps items 2`] = `
Object {
  "root.nodes.7.children": Array [
    6,
    3,
  ],
  "root.nodes.7.nodes.3": Object {
    "children": Array [
      1,
      2,
    ],
    "id": 3,
    "nodes": Object {
      "1": Object {
        "id": 1,
        "text": "1",
        "type": "plain-text",
      },
      "2": Object {
        "id": 2,
        "text": "a",
        "type": "plain-text",
      },
    },
    "props": Object {},
    "text": undefined,
    "type": "view",
  },
  "root.nodes.7.nodes.3.nodes.1": Object {
    "id": 1,
    "text": "1",
    "type": "plain-text",
  },
  "root.nodes.7.nodes.6.nodes.4": Object {
    "id": 4,
    "text": "0",
    "type": "plain-text",
  },
}
`;

exports[`wechat remax render swaps items 3`] = `
Object {
  "root.nodes.7.children": Array [
    3,
    6,
  ],
  "root.nodes.7.nodes.3.nodes.1": Object {
    "id": 1,
    "text": "0",
    "type": "plain-text",
  },
  "root.nodes.7.nodes.6": Object {
    "children": Array [
      4,
      5,
    ],
    "id": 6,
    "nodes": Object {
      "4": Object {
        "id": 4,
        "text": "1",
        "type": "plain-text",
      },
      "5": Object {
        "id": 5,
        "text": "b",
        "type": "plain-text",
      },
    },
    "props": Object {},
    "text": undefined,
    "type": "view",
  },
  "root.nodes.7.nodes.6.nodes.4": Object {
    "id": 4,
    "text": "1",
    "type": "plain-text",
  },
}
`;

exports[`wechat remax render umount component 1`] = `
Object {
  "children": Array [
    3,
  ],
  "id": 0,
  "nodes": Object {
    "3": Object {
      "children": Array [
        2,
      ],
      "id": 3,
      "nodes": Object {
        "2": Object {
          "children": Array [
            1,
          ],
          "id": 2,
          "nodes": Object {
            "1": Object {
              "id": 1,
              "text": "foo",
              "type": "plain-text",
            },
          },
          "props": Object {},
          "text": undefined,
          "type": "view",
        },
      },
      "props": Object {},
      "text": undefined,
      "type": "view",
    },
  },
  "props": undefined,
  "text": undefined,
  "type": "root",
}
`;

exports[`wechat remax render umount component 2`] = `
Object {
  "children": Array [
    3,
  ],
  "id": 0,
  "nodes": Object {
    "3": Object {
      "children": Array [],
      "id": 3,
      "props": Object {},
      "text": undefined,
      "type": "view",
    },
  },
  "props": undefined,
  "text": undefined,
  "type": "root",
}
`;
